torch/autograd/gradcheck.py

```diff
@@ -308,23 +321,65 @@ def fn(input):
                                                                                                 nondet_tol=nondet_tol)
         numerical = get_numerical_jacobian(fn, tupled_inputs, eps=eps)
 
-        if not correct_grad_sizes:
-            return fail_test('Analytical gradient has incorrect size')
+        out_is_complex = o.is_complex()
+
+        if out_is_complex:
+            # analytical vjp with v = 1.0j
+            analytical_with_imag_v, reentrant_with_imag_v, \
+                correct_grad_sizes_with_imag_v, correct_grad_types_with_imag_v \
+                = get_analytical_jacobian(tupled_inputs, o, nondet_tol=nondet_tol, v=1j)
+            numerical_with_imag_v = get_numerical_jacobian(fn, tupled_inputs, eps=eps, v=1j)
 
         if not correct_grad_types and check_grad_dtypes:
             return fail_test('Gradient has dtype mismatch')
 
-        for j, (a, n) in enumerate(zip(analytical, numerical)):
+        if out_is_complex and not correct_grad_types_with_imag_v and check_grad_dtypes:
+            return fail_test('Gradient (calculated using complex valued grad output) has dtype mismatch')
+
+        if not correct_grad_sizes:
+            return fail_test('Analytical gradient has incorrect size')
+
+        if out_is_complex and not correct_grad_sizes_with_imag_v:
+            return fail_test('Analytical gradient (calculated using complex valued grad output) has incorrect size')
+
+        def checkIfNumericalAnalyticAreClose(a, n, j, error_str=''):
+            if not torch.allclose(a, n, rtol, atol):
+                return fail_test(error_str + 'Jacobian mismatch for output %d with respect to input %d,\n'
+                                 'numerical:%s\nanalytical:%s\n' % (i, j, n, a))
+
+        inp_tensors = iter_tensors(tupled_inputs, True)
+
+        for j, (a, n, inp) in enumerate(zip(analytical, numerical, inp_tensors)):
             if a.numel() != 0 or n.numel() != 0:
-                if not torch.allclose(a, n, rtol, atol):
-                    return fail_test('Jacobian mismatch for output %d with respect to input %d,\n'
-                                     'numerical:%s\nanalytical:%s\n' % (i, j, n, a))
+                if o.is_complex():
+                    # C -> C, R -> C
+                    a_with_imag_v = analytical_with_imag_v[j]
+                    n_with_imag_v = numerical_with_imag_v[j]
+                    checkIfNumericalAnalyticAreClose(a_with_imag_v, n_with_imag_v, j,
+                                                     "Imaginary value(s) of Conjugate Wirtinger derivative \
```

Comment:
If this is just the imaginary part of the Conjugate Wirtinger, could you actually get, for the numerical jacobian, both this value and the one required line 363 below by taking the real and imaginary part of what is computed line 83: `conj_w_d = 0.5 * (ds_dx + ds_dy * 1j)` ?
This would remove one of the two calls to `get_numerical_jacobian`.

Reply:
The message mentioned earlier was incorrect. For `grad_out v = v1 + v2 j`,
``` 
ds_dJ_conj = 0.5 * [ v.conj() * (ds_dx + ds_dy * 1j) + v * (ds_dx - ds_dy * 1j) ]
                   = v1 * ds_dx.real + v2 * (ds_dx.imag + ds_dy.real) + v1 * (ds_dy.real + ds_dy.imag) * (1 + 1j)

For v = 1, we have v1 = 1, v2 = 0
ds_dJ_conj = ds_dx.real +(ds_dy.real + ds_dy.imag) * (1 + 1j)

For v = 1j, we have v1 = 0, v2 = 1
ds_dJ_conj = ds_dx.imag + ds_dy.real
```

The special cases clearly noticeable are:
```
- For  v=1, and f:`R->C`, ds_dJ_conj = ds_dx.real
- For v=1j, and f:`R->C`, ds_dJ_conj = 0
```